src/protocol/resp/src/parse/parser.rs

```diff
@@ -0,0 +1,434 @@
+        self.data = rest;
+        Ok(head)
+    }
+
+    fn parse_crlf(&mut self) -> ParseResult<'a, ()> {
+        self.parse_literal(b"\r\n").map(|_| ())
+    }
+
+    fn parse_len(&mut self) -> ParseResult<'a, usize> {
+        self.try_parse(|p| {
+            let text = p.parse_delimited_text()?;
+            let text = std::str::from_utf8(text).map_err(|_| ParseError::InvalidNumber(text))?;
+            text.parse()
+                .map_err(|_| ParseError::InvalidNumber(text.as_bytes()))
+        })
+    }
+
+    /// Parses text not containing CR or LF followed by CRLF
+    pub(crate) fn parse_delimited_text(&mut self) -> ParseResult<'a, &'a [u8]> {
+        self.try_parse(|p| {
+            let text = match memchr::memchr2(b'\r', b'\n', p.data) {
+                Some(offset) => p.parse_bytes(offset)?,
+                None => return Err(ParseError::Incomplete),
+            };
+
+            p.parse_crlf()?;
+
+            Ok(text)
+        })
+    }
+
+    pub fn parse_simple_string(&mut self) -> ParseResult<'a, &'a [u8]> {
+        self.try_parse(|p| {
+            p.parse_literal(b"+")?;
+            p.parse_delimited_text()
+        })
+    }
+
+    pub fn parse_bulk_string(&mut self) -> ParseResult<'a, Option<&'a [u8]>> {
+        self.try_parse(|p| {
+            p.parse_literal(b"$")?;
+
+            let result = if matches!(p.peek(), Some(b'-')) {
+                p.parse_literal(b"-1")?;
+                None
+            } else {
+                let len = p.parse_len()?;
+                let data = p.parse_bytes(len)?;
+                Some(data)
+            };
+
+            p.parse_crlf()?;
+            Ok(result)
+        })
+    }
+
+    pub fn parse_string(&mut self) -> ParseResult<'a, Option<&'a [u8]>> {
+        match self.peek() {
+            Some(b'+') => self.parse_simple_string().map(Some),
+            Some(b'$') => self.parse_bulk_string(),
+            Some(c) => Err(ParseError::UnexpectedCharacter {
+                expected: "\"+\" or \"$\"",
+                found: c,
+            }),
+            None => Err(ParseError::Incomplete),
+        }
+    }
+
+    pub fn parse_error(&mut self) -> ParseResult<'a, &'a [u8]> {
+        self.try_parse(|p| {
+            p.parse_literal(b"-")?;
+            p.parse_delimited_text()
+        })
+    }
+
+    pub fn parse_literal_u64(&mut self) -> ParseResult<'a, u64> {
+        self.try_parse(|p| {
+            p.parse_literal(b":")?;
+            let text = p.parse_delimited_text()?;
+            let text = std::str::from_utf8(text).map_err(|_| ParseError::InvalidNumber(text))?;
+            text.parse()
+                .map_err(|_| match <i64 as FromStr>::from_str(text) {
+                    Ok(_) => ParseError::UnexpectedNegativeNumber(text.as_bytes()),
+                    Err(_) => ParseError::InvalidNumber(text.as_bytes()),
+                })
+        })
+    }
+
+    pub fn parse_literal_i64(&mut self) -> ParseResult<'a, i64> {
```

Comment:
We should consider using the optimized i64 parsing I wrote for the `incr` command, or something better if we come up with it. Probably same with u64 but that doesn't exist yet. Mostly for coherence and predictability of parsing behavior, as we've established that a few nanoseconds is unlikely to matter much. Either that or I should modify the `incr` pr to use the stdlib parse.

Reply:
I expect that the std parsing is good enough that it is easier to just use it. I don't think it'll make enough of a difference in performance either way.